src/dialects/ibmi/query-generator.js

```diff
@@ -0,0 +1,836 @@
+    if (typeof value === 'boolean') {
+      return value.toString();
+    }
+
+    // null is stored as a string in mysql
+    if (value === null) {
+      return 'null';
+    }
+
+    return value;
+  }
+
+  upsertQuery(tableName, insertValues, updateValues, where, model, options) {
+    const aliasTable = `temp_${this.quoteTable(tableName)}`;
+
+    let query = `MERGE INTO ${this.quoteTable(tableName)} `;
+
+    const usingClause = `USING (
+      SELECT * FROM (${this.quoteTable(tableName)}
+      VALUES(42)
+      ) AS ${aliasTable}("id") ON (${aliasTable}."id" = ${this.quoteTable(tableName)}."id")`;
+
+    query += usingClause;
+    query += ` WHEN MATCHED THEN ${this.updateQuery(tableName, tableName, where, options, updateValues)}
+    WHEN NOT MATCHED THEN ${this.insertQuery(tableName, insertValues, model, options).sql}`;
+
+    return query;
+  }
+
+  insertQuery(table, valueHash, modelAttributes, options) {
+    // remove the final semi-colon
+    const query = super.insertQuery(table, valueHash, modelAttributes, options);
+    if (query.query[query.query.length - 1] === ';') {
+      query.query = query.query.slice(0, -1);
+      query.query = `SELECT * FROM FINAL TABLE (${query.query})`;
+    }
+
+    return query;
+  }
+
+  selectQuery(tableName, options, model) {
+    // remove the final semi-colon
+    let query = super.selectQuery(tableName, options, model);
+    if (query[query.length - 1] === ';') {
+      query = query.slice(0, -1);
+    }
+
+    return query;
+  }
+
+  bulkInsertQuery(tableName, fieldValueHashes, options, fieldMappedAttributes) {
+    // remove the final semi-colon
+    let query = super.bulkInsertQuery(tableName, fieldValueHashes, options, fieldMappedAttributes);
+    if (query[query.length - 1] === ';') {
+      query = query.slice(0, -1);
+      query = `SELECT * FROM FINAL TABLE (${query})`;
+    }
+
+    return query;
+  }
+
+  truncateTableQuery(tableName) {
+    return `TRUNCATE TABLE ${this.quoteTable(tableName)} IMMEDIATE`;
+  }
+
+  deleteQuery(tableName, where, options = {}, model) {
+    let limit = '';
+    if (options.offset || options.limit) {
+      limit = this.addLimitAndOffset(options);
+    }
+
+    let query = `DELETE FROM ${this.quoteTable(tableName)}`;
+
+    where = this.getWhereConditions(where, null, model, options);
+
+    if (where) {
+      query += ` WHERE ${where}`;
+    }
+
+    return query + limit;
+  }
+
+  /**
+   * Returns an SQL fragment for adding result constraints.
+   *
+   * @param  {object} options An object with selectQuery options.
+   * @returns {string}         The generated sql query.
+   * @private
+   */
+  addLimitAndOffset(options) {
+    let fragment = '';
+    const offset = options.offset;
+    const limit = options.limit;
+
+    if (offset) {
+      if (typeof offset === 'number' && Number.isInteger(offset)) {
+        fragment += ` OFFSET ${offset} ROWS`;
+      } else {
+        console.warn('"offset" must be an integer. Offset is not added');
```

Comment:
I would throw here instead

Reply:
Will cause https://github.com/sequelize/sequelize/blob/main/test/unit/sql/offset-limit.test.js#L76 to fail. Every other dialect (excepting `db2`) seems to just accept the input, rather than even do a type check.

Reply:
Thanks :)
I think that's something we should look into changing in v7 but we can do that in another PR